### PR TITLE
Reintegrating scabha, and also #47

### LINKED ISSUE #32: Fix pytests
I've gone through quite a bit of this already, but better for you to fix it @o-smirnov. 

see [the errors here](https://github.com/caracal-pipeline/stimela2/runs/5840302057?check_suite_focus=true) and commit to branch `fix-pytests`

### LINKED ISSUE #47: custom runners selected by command prefix
Thinking of #46, I thought let's have a separate bit of code that

* takes a cab's parameters

* converts them into a command line using this 

    ```yml
    command: casa --nogui --log2term -c 'flagmanager(vis="{current.ms}", mode="{current.mode}", versionname="{current.name}")'
    ```
for inspiration. So then we can have a have a "CASA-flavour" cab which is different from a "normal command cab" in the sense that the command-line is assembled differently (and perhaps failure checking, and post-run cleanup is different). And then for every required CASA task we just define a separate cab where we specify the task name, and the inputs/outputs, and the rest is standard (and uses the same CASA base image).

The I realized we already have a this distinction being made here: https://github.com/caracal-pipeline/stimela2/blob/master/stimela/backends/native.py#L33. We have the normal command cabs, and we have Python callable cabs. The latter is distinguished by ``(module_name)function_name``.

So, I propose we have a general scheme for specifying the flavour via the command string:

```yml
command: command_name       # regulation cab
command: (module)function       # Python callable cab
command: casa::taskname       # CASA-flavour cab 
command: python::(module)function       # alternative syntax for Python callable cab (for consistency)
```

...which also allows for future "flavours"...

### PR DESCRIPTION
**Pull request by o-smirnov**
@SpheMakh don't merge yet. Opening PR so that tests can start running.

### CODE REVIEW

File: stimela/commands/clean.py [hunk 1]
```diff
@@ -25,59 +25,61 @@ def make_parser(subparsers):
...
-                # Kill the container instead of stopping it, so that effect can be felt py parent process
-                utils.xrun('docker', ['kill', container])
-                cont.remove()
-            elif status in ['exited', 'dead']:
-                cont.remove()
-
-            log.remove('containers', container)
-            log.write()
+    pass
+
```
Comment (sphemakh) [inline stimela/commands/clean.py]: Let's delete this old code instead of commenting it out

File: stimela/commands/kill.py [hunk 2]
```diff
@@ -13,22 +13,23 @@ def make_parser(subparsers):
 
 
 def kill(args, conf):
-    log = logger.StimelaLogger(LOG_FILE)
+    pass
+    # log = logger.StimelaLogger(LOG_FILE)
```
Comment (sphemakh) [inline stimela/commands/kill.py]: ditto

File: stimela/commands/pull.py [hunk 3]
```diff
@@ -35,83 +35,86 @@ def make_parser(subparsers):
...
-                log.log_image(image, 'pulled')
-            elif args.podman:
-                podman.pull(image, force=args.force)
-                log.log_image(image, 'pulled')
-            else:
-                docker.pull(image, force=args.force)
-                log.log_image(image, 'pulled')
-
-    log.write()
+    pass
```
Comment (sphemakh) [inline stimela/commands/pull.py:38]: ditto

File: tests/scabha_tests/test_schema.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    OmegaConf.save(conf, sys.stdout)
+
+    print(type(conf.test.a.default))
+
+    obj = OmegaConf.to_object(conf)
+    print(obj)
+
+
+if __name__ == "__main__":
+    test_schema()
```
Comment (sphemakh) [inline tests/scabha_tests/test_schema.py]: add newline

File: tests/scabha_tests/test_schema.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+test:
+  a:
+    dtype: List[int]
+    default: [1,2]
```
Comment (sphemakh) [inline tests/scabha_tests/test_schema.yaml]: add newline

File: tests/scabha_tests/test_substitutions.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    try:
+        context.evaluate("xxx")
+        raise RuntimeError("exception should have been raised due to invalid substitution")
+    except SubstitutionError as exc:
+        print(f"Error as expected ({exc})")
+
+
+    # # test <<-substitutions
+    
+    # with substitutions_from(ns) as context:
```
Comment (sphemakh) [inline tests/scabha_tests/test_substitutions.py]: nuke old code

**Review by SpheMakh — CHANGES_REQUESTED** [review]

**Comment by o-smirnov** [comment]
OK it's all green and I addressed your review. @SpheMakh will you do the merge honours please.

**Comment by o-smirnov** [comment]
@SpheMakh reminder.

**Review by SpheMakh — APPROVED** [review]

**Pull request merged**